openupgrade_scripts/scripts/sale_loyalty/16.0.1.0/pre-migration.py

```diff
@@ -0,0 +1,299 @@
+        WHERE sol.product_id = lr.discount_line_product_id
+        """,
+    )
+    # Fill coupon_id field(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE sale_order_line
+        ADD COLUMN IF NOT EXISTS coupon_id INT
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE sale_order_line
+        SET coupon_id = loyalty_card.id
+        FROM loyalty_card
+        WHERE sale_order_line.order_id = loyalty_card.sales_order_id
+        """,
+    )
+    # Fill reward_identifier_code
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE sale_order_line
+        ADD COLUMN IF NOT EXISTS reward_identifier_code VARCHAR
+        """,
+    )
+    reward_code = _generate_random_reward_code()
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE sale_order_line
+        SET reward_identifier_code = %s
+        WHERE reward_id IS NOT NULL
+        """,
+        (reward_code,),
+    )
+    # Fill points_cost field
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE sale_order_line
+        ADD COLUMN IF NOT EXISTS points_cost FLOAT
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE sale_order_line
+        SET points_cost = ABS(untaxed_amount_to_invoice)
+        WHERE reward_id IS NOT NULL
+        AND untaxed_amount_to_invoice < 0
+        """,
+    )
+
+
+def delete_sql_constraints(env):
+    # Delete constraints to recreate it
+    openupgrade.delete_sql_constraint_safely(
+        env, "sale_loyalty", "sale_order_coupon_points", "order_coupon_unique"
+    )
+
+
+def update_template_keys(env):
+    """Update template keys of the merged sale_gift_card module in loyalty_sale"""
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE ir_ui_view
+        SET key = 'sale_loyalty.used_gift_card'
+        WHERE key = 'sale_gif_card.used_gift_card'
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE ir_ui_view
+        SET key = 'sale_loyalty.sale_purchased_gift_card'
+        WHERE key = 'sale_gif_card.sale_purchased_gift_card'
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE ir_ui_view
+        SET key = 'sale_loyalty.sale_orde_portal_content_inherit'
+        WHERE key = 'sale_gif_card.sale_orde_portal_content_inherit'
+        """,
+    )
+
+
+def merge_sale_gift_card_to_sale_loyalty_card(env):
+    # Update the coupon_id column in the sale_order_line table with the ID of the
+    # loyalty_card table based on certain criteria and relationships established between
+    # the loyalty_card, loyalty_reward and gift_card tables
+    # program_id is added to the gift_card table in the loyalty migration script.
+    table = openupgrade.get_model2table("gift.card")
+    if not openupgrade.table_exists(env.cr, table):
```

Comment:
No need to do workarounds, as we know the table name:

```suggestion
    table = openupgrade.get_model2table("gift.card")
    if not openupgrade.table_exists(env.cr, "gift_card"):
```